e2e_context_test.go

```diff
@@ -393,3 +393,160 @@ func TestContextPerRequestWithTimeoutCancelledByContext(t *testing.T) {
 	// expected error should occur
 	assert.True(t, suppressor.expErrorOccurred)
 }
+
+func TestContextPerRequestRetry(t *testing.T) {
+	var m sync.Mutex
+	TimeOutDuration := 5 * time.Minute
+
+	t.Run("not cancelled", func(t *testing.T) {
+		var callCount int
+		var isCtxCancelled bool
+		ctxCancellationSleepDuration := 1 * time.Minute
+		retryDelayDuration := 1 * time.Second
+
+		ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			m.Lock()
+			defer m.Unlock()
+			callCount++
+
+			if callCount > 1 {
+				w.WriteHeader(http.StatusOK)
+				return
+			}
+
+			w.WriteHeader(http.StatusInternalServerError)
+		}))
+		defer ts.Close()
+
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		go func() {
+			time.Sleep(ctxCancellationSleepDuration)
+			cancel()
+			m.Lock()
+			defer m.Unlock()
+			isCtxCancelled = true
+		}()
+
+		// Config with all error
+		suppressor := newExpErrorSuppressor(t,
+			func(err error) bool {
+				return true
+			})
+
+		e := WithConfig(Config{
+			BaseURL:          ts.URL,
+			AssertionHandler: suppressor,
+		})
+
+		e.GET("/").
+			WithContext(ctx).
+			WithMaxRetries(1).
+			WithRetryPolicy(RetryAllErrors).
+			WithRetryDelay(retryDelayDuration, TimeOutDuration).
+			Expect()
+
+		assert.Equal(t, 2, callCount)
+		assert.False(t, isCtxCancelled)
+		assert.False(t, suppressor.expErrorOccurred)
+	})
+
+	t.Run("cancelled on first retry attempt", func(t *testing.T) {
+		var callCount int
+		var isCtxCancelled bool
+		ctxCancellationSleepDuration := 1 * time.Second
+		retryDelayDuration := 1 * time.Minute
+
+		ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			m.Lock()
+			defer m.Unlock()
+			callCount++
+			w.WriteHeader(http.StatusInternalServerError)
+		}))
+		defer ts.Close()
+
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		go func() {
+			time.Sleep(ctxCancellationSleepDuration)
+			cancel()
+			m.Lock()
+			defer m.Unlock()
+			isCtxCancelled = true
+		}()
+
+		// Config with context cancelled error
+		suppressor := newExpErrorSuppressor(t,
+			func(err error) bool {
+				return strings.Contains(err.Error(), context.Canceled.Error())
+			})
+
+		e := WithConfig(Config{
+			BaseURL:          ts.URL,
+			AssertionHandler: suppressor,
+		})
+
+		e.GET("/").
+			WithContext(ctx).
+			WithMaxRetries(1).
+			WithRetryPolicy(RetryAllErrors).
+			WithRetryDelay(retryDelayDuration, TimeOutDuration).
+			Expect()
+
+		assert.Equal(t, 1, callCount)
+		assert.True(t, isCtxCancelled)
+		assert.True(t, suppressor.expErrorOccurred)
```

Comment:
Here you assume that cancel() will be called after first retry and before second retry. This will usually work, however it's not guaranteed, and this test is not deterministic. Examples:

* if httptest goroutine is scheduled lately, cancel() may be called before first try
* if cancellation goroutine is scheduled lately after sleep expires, cancel() may be called after second try

And so on. You can't rely on specific scheduling order.

To keep things simple and deterministic at the same time, I suggest the following:

* merge "cancelled on first retry attempt" and "cancelled on second retry attempt" into one test
* don't assume that cancel is called after first and before second try; only assume that it's called after first try
* to guarantee that it's called only after first try, use ctxCancellation channel like you did in "cancelled on second retry attempt"
* set WithMaxRetries to large value (don't limit retries)
* set min and max in WithRetryDelay to very small values (a few ms) to keep tests fast

Reply:
got it will do the update